### DISCUSSION TITLE
Battery test based on a given logic and parameterize them by fixtures

### DISCUSSION THREAD

**Discussion by jaraqueffdc** [pytest-dev/pytest#8623] [Q&A]
Hi!

I have worked with pytest in the past and I loved it but for one reason or another all the project I have worked in use unittest but I found an use case where I cannot use unittest and I was hopping pytest would allow me to use it. 

I am sorry if this is simple to do but I think the issue is that I do not know what to look for exactly 😅

The scenario is the following:
* I have an abstract class that defines most of my logic but there are certain specifications that depend on the derived class
* I want to define the entire battery of tests based on the public interface which depends only on the public interface, implemented in the abstract class itself. If not, I would be repeating the same tests for the different specifications.

I have all the tests defined and I would like to be able to use those but being able to configure each set of tests because we have 3 different specifications. All tests depends on a fixture `ds` which is the specification itself but for some of them we will also need mocking which I was planning to do with a yielding fixture with autouse.

However, I am not sure how to separate the three test sets so each fixture is used with the written tests as well as allowing the mocking fixtures to only apply when running for that given specification. 

I am not sure it is clear but please let me know if you have any question about the intended use or any suggestion.

Thanks a lot!

**Comment by portyu9** [comment]
Hi @jaraqueffdc 

A parametrized fixture is a good fit for this.

Put the implementations/specifications in the fixture parameters:

```python
@pytest.fixture(
    params=[SpecA, SpecB, SpecC],
    ids=lambda spec: spec.__name__,
)
def spec(request):
    return request.param()
```

Then write the common battery only once:

```python
def test_feature_one(spec):
    assert spec.feature_one()


def test_feature_two(spec):
    assert spec.feature_two()
```

pytest will run each test once for every fixture parameter.

If an implementation needs additional mocks/setup, another fixture can depend on `spec`:

```python
@pytest.fixture
def prepared_spec(spec, monkeypatch):
    if isinstance(spec, SpecB):
        # SpecB-specific setup
        ...

    return spec
```

I would generally keep that dependency explicit rather than making it autouse unless **every** test in its scope should automatically become parametrized by every specification.

Please mark this answer as accepted if it helped, thank you.
